jax/numpy/lax_numpy.py

```diff
@@ -2807,8 +2810,10 @@ def filter_singleton_dims(operand, names, other_shape, other_names):
       rhs, rhs_names = sum_repeats(rhs, rhs_names, rhs_counts,
                                    result_names + lhs_names)
 
-      contracted_names = contracted_names & (set(lhs_names) | set(rhs_names))
-      batch_names = (set(lhs_names) & set(rhs_names)) - contracted_names
+      contracted_names = [x for x in contracted_names
+                          if x in set(lhs_names) or x in set(rhs_names)]
```

Comment:
These numbers are probably small enough that it doesn't matter, but you could create the sets outside the list comprehension (I'm assuming it creates new sets on every predicate evaluation). Alternatively, keep everything as sets like before and convert to a sorted list after this line. (Or leave as-is)